File: estructurales/bridge/BridgeExample.java
interface Device {
    boolean isOn();
    void powerOn();
    void powerOff();
}

class TV implements Device {
    private boolean on = false;

    @Override
    public boolean isOn() { return on; }

    @Override
    public void powerOn() {
        on = true;
        System.out.println("TV: ON");
    }

    @Override
    public void powerOff() {
        on = false;
        System.out.println("TV: OFF");
    }
}

class Radio implements Device {
    private boolean on = false;

    @Override
    public boolean isOn() { return on; }

    @Override
    public void powerOn() {
        on = true;
        System.out.println("Radio: ON");
    }

    @Override
    public void powerOff() {
        on = false;
        System.out.println("Radio: OFF");
    }
}

// Abstraction: se conecta a Device (bridge)
class RemoteControl {
    protected final Device device;

    public RemoteControl(Device device) {
        this.device = device;
    }

    public void togglePower() {
        if (device.isOn()) device.powerOff();
        else device.powerOn();
    }
}

public class BridgeExample {
    public static void main(String[] args) {
        RemoteControl remoteTV = new RemoteControl(new TV());
        remoteTV.togglePower();
        remoteTV.togglePower();

        RemoteControl remoteRadio = new RemoteControl(new Radio());
        remoteRadio.togglePower();
    }
}

File: estructurales/bridge/BridgeExample.py
class Device:
    """Implementor"""
    def is_on(self) -> bool:
        raise NotImplementedError

    def power_on(self) -> None:
        raise NotImplementedError

    def power_off(self) -> None:
        raise NotImplementedError


class TV(Device):
    def __init__(self) -> None:
        self._on = False

    def is_on(self) -> bool:
        return self._on

    def power_on(self) -> None:
        self._on = True
        print("TV: ON")

    def power_off(self) -> None:
        self._on = False
        print("TV: OFF")


class Radio(Device):
    def __init__(self) -> None:
        self._on = False

    def is_on(self) -> bool:
        return self._on

    def power_on(self) -> None:
        self._on = True
        print("Radio: ON")

    def power_off(self) -> None:
        self._on = False
        print("Radio: OFF")


class RemoteControl:
    """Abstraction (bridge to Device)"""
    def __init__(self, device: Device) -> None:
        self.device = device

    def toggle_power(self) -> None:
        if self.device.is_on():
            self.device.power_off()
        else:
            self.device.power_on()


if __name__ == "__main__":
    remote_tv = RemoteControl(TV())
    remote_tv.toggle_power()
    remote_tv.toggle_power()

    remote_radio = RemoteControl(Radio())
    remote_radio.toggle_power()